src/socket/hackney_pool.erl

```diff
@@ -1,437 +1,286 @@
-%% @doc get timeout
-timeout(Name) ->
-    gen_server:call(find_pool(Name), timeout).
 
-%% @doc change the connection timeout
-%%
-set_timeout(Name, NewTimeout) ->
-    gen_server:cast(find_pool(Name), {set_timeout, NewTimeout}).
+reuse_connection(Group, State) ->
+    Refs = ets:lookup(State#state.refs, {conns, Group}),
+    reuse_connection1(Refs, State).
 
-%% @private
-%%
-%%
-do_start_pool(Name, Options) ->
-    Spec = child_spec(Name, Options),
-    case supervisor:start_child(hackney_sup, Spec) of
-        {ok, Pid} ->
-            Pid;
-         {error, {already_started, _}} ->
-            find_pool(Name, Options)
-    end.
 
+reuse_connection1([], _State) ->
+    no_socket;
+reuse_connection1([{Group, #hackney_socket{sock=S}=HS} | Rest], State) ->
+    [{_, _, HS, T}] = ets:lookup(State#state.sockets, S),
 
-find_pool(Name) ->
-    case ets:lookup(?MODULE, Name) of
-        [] ->
-            undefined;
-        [{_, Pid}] ->
-            Pid
+    ets:delete_object(State#state.refs, {Group, HS}),
+    ets:delete(State#state.sockets, S),
+    cancel_timer(T, S),
+    hackney_socket:setopts(HS, [{active, false}]),
+    case sync_socket(HS) of
+        true ->
+            {ok, HS};
+        false ->
+            reuse_connection1(Rest, State)
     end.
 
-find_pool(Name, Options) ->
-     case ets:lookup(?MODULE, Name) of
-        [] ->
-            do_start_pool(Name, Options);
-        [{_, Pid}] ->
-            Pid
+request_socket(Group, Req, State) ->
+    %% we simply balance connections tasks between connectors using an RR algorithm
+    {Connector, State2} = pick_connector(State),
+    Connector ! {connect, Group, Req},
+    State2.
+
+
+pick_connector(State=#state{connectors=Connectors}) ->
+    [Connector | Rest] = Connectors,
+    {Connector, State#state{connectors = Rest ++ [Connector]}}.
+
+
+release_socket(#hackney_socket{group=Group}=HS, State) ->
+    Pending = ets:lookup(State#state.refs, {pending, Group}),
+    case Pending of
+        [] -> cache_socket(HS, State);
+        _ -> dispatch_socket(Pending, HS, State)
     end.
 
 
-start_link(Name, Options0) ->
-    Options = hackney_util:maybe_apply_defaults([max_connections, timeout],
-                                                Options0),
-    gen_server:start_link(?MODULE, [Name, Options], []).
+cache_socket(HS=#hackney_socket{sock=Sock, group=Group}, State) ->
+    Conns = ets:lookup(State#state.refs, {conns, Group}),
+    TotalGroup = length(Conns),
+    TotalConns = ets:info(State#state.sockets, size),
 
-init([Name, Options]) ->
-    process_flag(priority, high),
-    case lists:member({seed,1}, ssl:module_info(exports)) of
-        true ->
-            % Make sure that the ssl random number generator is seeded
-            % This was new in R13 (ssl-3.10.1 in R13B vs. ssl-3.10.0 in R12B-5)
-            apply(ssl, seed, [crypto:rand_bytes(255)]);
-        false ->
-            ok
+    Limit = case Group of
+            <<"proxy/">> -> State#state.proxy_limit;
```

Comment:
it should match the suffix there. ie. `<<"proxy/", _/binary>>` . also it should match the socks group.

> note: itmay be better instead to pass different limits matching some pattern as an option instead, so different groups limits can directly be handled.